flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestServerEndpoint.java

```diff
@@ -181,14 +184,44 @@ protected void initChannel(SocketChannel ch) {
 				.channel(NioServerSocketChannel.class)
 				.childHandler(initializer);
 
-			log.debug("Binding rest endpoint to {}:{}.", restBindAddress, restBindPort);
-			final ChannelFuture channel;
-			if (restBindAddress == null) {
-				channel = bootstrap.bind(restBindPort);
-			} else {
-				channel = bootstrap.bind(restBindAddress, restBindPort);
+			ChannelFuture channel;
+
+			// parse port range definition and create port iterator
+			Iterator<Integer> portsIterator;
+			try {
+				portsIterator = NetUtils.getPortRangeFromString(restBindPort);
+			} catch (Exception e) {
+				throw new IllegalArgumentException("Invalid port range definition: " + restBindPort);
+			}
+
+			int chosenPort = 0;
+			while (portsIterator.hasNext()) {
+				try {
+					chosenPort = portsIterator.next();
+					if (restBindAddress == null) {
+						channel = bootstrap.bind(chosenPort);
+					} else {
+						channel = bootstrap.bind(restBindAddress, chosenPort);
+					}
+					serverChannel = channel.syncUninterruptibly().channel();
+					break;
+				} catch (Exception e) {
+					// we can continue to try if this contains a netty channel exception
+					Throwable cause = e.getCause();
+					if (!(cause instanceof org.jboss.netty.channel.ChannelException ||
+						cause instanceof java.net.BindException)) {
+						throw e;
```

Comment:
Here, we just retry when the exception is a port conflict specific exception. If not, it seems we should throw it and let the caller know there is another exception. Also, see `BootstrapTools#startActorSystem`.

Reply:
Yes, this expression may be misunderstood, I will update it later. Regarding the exception handling logic here, in the recent test, it is really a problem, we should not call it again.

```
Throwable cause = e.getCause();
```

It will get a `null` value. I will fix it later.